### ISSUE TITLE
Clarification of normal direction (SDF direction) and Sharp edge sampling process (mistake?)

### ISSUE THREAD

**Issue by jseobyun** [stepfun-ai/Step1X-3D#38]
Hello authors! I truly appreciate for you kind response. Thank you so much.

After you updated the data preprocessing script (watertight_and_sampling.py), I tried constructing my own dataset using it. However, I ran into a couple of points that I found a bit confusing and would appreciate your clarification.

**1. First thing is reversed normal direction.** 

Following your code as-is, the exported mesh seems to have reversed normals, as shown below:

<img width="320" height="320" alt="Image" src="https://github.com/user-attachments/assets/776117c2-d27c-441e-8b56-81b53ba0712d" /> <img width="410" height="380" alt="Image" src="https://github.com/user-attachments/assets/5b513335-3222-48dc-98ff-692e92a30d89" />
The normal direction appears flipped in multiple mesh viewers, so I believe it's not viewer-specific.

Is this inversion of normals intentional in your pipeline? If so, could you please elaborate on the reasoning? Otherwise, wouldn't it be more appropriate to follow the usual convention by negating the SDF values at the following point in the code?

What is your SDF definition in your pipeline? (negative : inside, positive : outside?)

```
## in def watertight
## STEP D. generate the mesh using Marching Cube
sdf = dist.view(grid_resolution, grid_resolution, grid_resolution) * -1 # here -1 added
```
Your clarification would help a lot!

**2. Second thing is sharp edge sampling from input mesh** 

At the end of the script, you sample sharp edge points from args.input_mesh:
```
print("Sampling points on sharp edges of the mesh.")
sharp_surface = sharp_edge_sampling(
        args.input_mesh, 
        num_views=args.num_sharp_surface_points,
        sharpness_threshold=math.radians(args.angle_threshold)
)
```

However, since args.input_mesh is used before scaling and translation, the sampled points are not aligned with the normalized mesh (mesh.vertices).

This leads to a visible mismatch between the mesh and the sampled points, as shown below:

<img width="410" height="380" alt="Image" src="https://github.com/user-attachments/assets/cd3f28f7-070d-486a-ad9a-f93c1f70c977" />

Wouldn't it be more consistent to perform sharp edge sampling after mesh normalization, using the transformed mesh instead of args.input_mesh?

Again, thank you for your work and for sharing the code. I love your codes!

**Comment by jseobyun** [comment]
Oh, I found out that mcubes assume (outside : negative, inside :positive) , so only the result mesh has reversed normal. 

f = SDF(mesh.vertices, mesh.faces) with mcubes output works correctly. 

As a result before exporting the mesh, I think flipping normals required.  ( mesh.invert() )

**Comment by jseobyun** [comment]
I found that the normal flipping issue implies that there are multiple sign (+-) issues in the script.

**1. Your implementation assumes SDF (inside :-, outside : +)**

```
## STEP D. generate the mesh using Marching Cube
sdf = dist.view(grid_resolution, grid_resolution, grid_resolution) 
```
The output sdf of above part correctly follows the convention

<img width="470" height="425" alt="Image" src="https://github.com/user-attachments/assets/15e951c6-067e-454d-9976-d2880961b379" />

(positive :green, negative : red)



**2. mcubes marching cube assumes reversed SDF (inside : +, outside : -)**  <- reversed!

That's why the mesh has flipped normal, so mesh should be reverse face winding via mesh.invert(). 

mesh.invert() makes sure that saved normals in .npz are toward outside like this. 

<img width="500" height="420" alt="Image" src="https://github.com/user-attachments/assets/9f13187a-801b-4662-9ee1-62e2192a4100" />

(before mesh.invert())

<img width="500" height="420" alt="Image" src="https://github.com/user-attachments/assets/52d00822-cfdb-4a91-bdfe-3923f613933b" />

(after mesh.invert())


SDF ( inside: negative, outside :positive) -> mcubes.marching cubes -> flipped mesh -> mesh.invert() -> correct mesh

**3. pysdf SDF asuumes reversed SDF (inside : positive, outside :negative)**  <-reversed

Regardless of mesh's face winding (for example, mesh with/without mesh.invert()), pysdf output SDF (inside : +, outside : -) 

<img width="515" height="458" alt="Image" src="https://github.com/user-attachments/assets/2e3956ac-e895-4283-8190-5ee617a4b8f8" />


(positive :green, negative : red)


-----
As a summary, 
1. You assumes SDF(inside : -, outside : +) (correct!)
2. However, the outputs (mesh and saved SDFs) are saved assuming SDF(inside: +, outside -)

I'm confused whether you use SDF(inside : +, outside: -)  or SDF(inside : -, outside: +) in training data

Because I've tried to finetune provided model, clarifying the sign is important.

Please help me~

**Comment by haoz19** [comment]
Yes, I find the same problem, please help! Thanks so much!

**Comment by jseobyun** [comment]
I recently noticed that the "sharp_surface" array (N x 6) occasionally contains NaN values in the normal components.

Although I have reviewed the relevant code, I am still unsure why these NaNs occur. The number of such cases is small (typically fewer than 10), but they do appear intermittently.

For this reason, I would kindly suggest applying a filter to remove any rows containing NaN values before saving the .npz file.

```python
def remove_invalid_rows(name, array):
    if not isinstance(array, np.ndarray):
        print(f"{name} is not a valid numpy array.")
        return array

    invalid_mask = np.isnan(array) | np.isinf(array)
    row_has_invalid = np.any(invalid_mask, axis=1)

    if np.any(row_has_invalid):
        print(f"[{name}] Removed {np.sum(row_has_invalid)} invalid rows.")
        return array[~row_has_invalid]
    else:
        return array

data = np.load(surface_path) # N 6

cleaned_data = {
        "surface": remove_invalid_rows("surface", np.asarray(data["surface"])),
        "sharp_surface": remove_invalid_rows("sharp_surface", np.asarray(data["sharp_surface"])),
        "near_surface_points": remove_invalid_rows("near_surface_points", np.asarray(data["near_surface_points"])),
        "volume_rand_points": remove_invalid_rows("volume_rand_points", np.asarray(data["volume_rand_points"])),
        "bounds" : data["bounds"],
}
np.savez(surface_path, **cleaned_data)
```
